x-pack/platform/plugins/shared/observability_ai_assistant/server/service/distributed_lock_manager/index.ts

```diff
@@ -0,0 +1,329 @@
+              }
+            `,
+          params: {
+            token: this.token,
+            ttl,
+            metadata: options.metadata ?? {},
+          },
+        },
+        // @ts-expect-error
+        upsert: {},
+        refresh: true,
+      });
+
+      if (response.result === 'created') {
+        this.logger.debug(
+          `Lock "${this.lockId}" acquired with ttl = ${prettyMilliseconds(ttl)} and token = ${
+            this.token
+          }`
+        );
+        return true;
+      } else if (response.result === 'updated') {
+        this.logger.debug(
+          ` Lock "${this.lockId}" was expired and re-acquired with ttl = ${prettyMilliseconds(
+            ttl
+          )} and token = ${this.token}`
+        );
+        return true;
+      } else if (response.result === 'noop') {
+        this.logger.debug(
+          `Lock "${this.lockId}" could not be acquired with token ${this.token} because it is already held`
+        );
+        return false;
+      } else {
+        throw new Error(`Unexpected response: ${response.result}`);
+      }
+    } catch (e) {
+      if (
+        e instanceof errors.ResponseError &&
+        e.body?.error?.type === 'version_conflict_engine_exception'
+      ) {
+        this.logger.debug(`Lock "${this.lockId}" already held (version conflict)`);
+        return false;
+      }
+
+      this.logger.error(`Failed to acquire lock "${this.lockId}": ${e.message}`);
+      this.logger.debug(e);
+      return false;
+    }
+  }
+
+  /**
+   * Releases the lock by deleting the document with the given lockId and token
+   */
+  public async release(): Promise<boolean> {
+    try {
+      const res = await pRetry(
+        () => {
+          return this.esClient.deleteByQuery({
+            index: LOCKS_CONCRETE_INDEX_NAME,
+            query: {
+              // bool: { filter: [{ term: { _id: this.lockId } }, { term: { token: this.token } }] },
+              bool: { filter: [{ term: { _id: this.lockId } }, { term: { token: this.token } }] },
+            },
+            refresh: true,
+          });
+        },
+        { retries: 2 }
+      );
+
+      if (res.deleted === 0) {
+        this.logger.debug(
+          `Lock "${this.lockId}" with token = ${this.token} could not be released. Not found or already released.`
+        );
+        return false;
+      }
+
+      this.logger.debug(`Lock "${this.lockId}" with token = ${this.token} was released.`);
+      return true;
+    } catch (error) {
+      this.logger.error(`Failed to release lock "${this.lockId}": ${error.message}`);
+      this.logger.debug(error);
+      return false;
+    }
+  }
+
+  /**
+   * Releases a lock if it is expired.
+   */
+  public async releaseIfExpired(): Promise<boolean> {
+    const res = await this.esClient.deleteByQuery({
+      index: LOCKS_CONCRETE_INDEX_NAME,
+      query: {
+        bool: {
+          filter: [{ term: { _id: this.lockId } }, { range: { expiresAt: { lte: 'now' } } }],
```

Comment:
we shouldn't do this, as it opens us up for clock skew

Reply:
actually, I'm full of 💩, this is calculated on the ES side. But we should double check - if it's rewritten on the coordinating node, and not on the primary shard, there's still a chance of clock skew.

Reply:
How would you double check this? In Kibana? 

`releaseIfExpired` is only used by `get` which is not used anywhere at the moment. It's meant to help users lookup if a lock is active.

Reply:
I removed `releaseIfExpired`